dragonboard/io.py

```diff
@@ -193,17 +108,226 @@ def previous(self):
+    header_size = 4 * 16
+
+    EventHeader = namedtuple('EventHeader', [
+    'event_counter',
+    'trigger_counter',
+    'counter_133MHz',
+    'counter_10MHz',
+    'pps_counter',
+    'stop_cells',
+    'flag',
+    ])
+
+    def get_roi(self):
+        body_size = self.event_size - self.header_size
+        roi = body_size/(adc_word_size * num_gains * num_channels)
+        assert roi.is_integer()
+        return int(roi)
+
+    def read_header(self):
+        ''' return EventHeader from file f
+        '''
+        f = self.file_descriptor
+        (
+            header_aaaa, # should be constant 0xaaaa
+            pps_counter, # pps: pulse per second = 1Hz
+            counter_10MHz,
+            event_counter,
+            trigger_counter,
+            counter_133MHz,
+            data_header_all_ds, # should be constant 8 times 0xdd
+            flags,
+        ) = struct.unpack('!HHIIIQQ16s', f.read(struct.calcsize('!HHIIIQQ16s')))
+
+        assert header_aaaa == 0xaaaa, \
+            "Header is not 0xaaaa but {!r}".format(header_aaaa)
+        assert data_header_all_ds == 0xdddddddddddddddd, \
+            "data_header is not 8x 0xdd but {!r}".format(data_header_all_ds)
+
+        stop_cells_for_user = np.empty(
+            num_channels, dtype=[('low', 'i2'), ('high', 'i2')]
+        )
+        stop_cells__in_drs4_chip_order = np.frombuffer(
+            f.read(stop_cell_size), dtype=stop_cell_dtype)
+
+        for g,p in stop_cell_map:
+            chip = stop_cell_map[(g,p)]
+            stop_cells_for_user[g][p] = stop_cells__in_drs4_chip_order[chip]
+
+        return self.EventHeader(
+            event_counter, 
+            trigger_counter, 
+            counter_133MHz, 
+            counter_10MHz, 
+            pps_counter,
+            stop_cells_for_user,
+            flags
+        )
+
+    def guess_event_size(self):
+        ''' try to find out the event size for this file.
+
+        Each even header contains a data_header
+        The distance between two data_headers is just the event size.
+        '''
+        f = self.file_descriptor
+        current_position = f.tell()
+        f.seek(0)
+
+        chunk = self.get_chunk()
+
+        data_header = b'\xdd'*8
+        first_data_header = chunk.find(data_header)
+        second_data_header = chunk.find(data_header, first_data_header + 1)
+
+        event_size = second_data_header - first_data_header
+        f.seek(current_position)
+
+        return event_size
+
+version_map = {
+    "v5_1_05": EventGenerator_v5_1_05,
+    "v5_1_0B": EventGenerator_v5_1_0B,
+}
+
+def EventGenerator(path, max_events=None, version=None):
+    if version is None:
+        return version_map[guess_version(path)](path, max_events)
+    else:
+        return version_map[version](path, max_events)
+
+def guess_version(path):
+    for version_name in version_map:
+        EG = version_map[version_name]
+        try:
+            EG(path)
+            return version_name
+        except:
```

Comment:
I guess not, but I have currently no idea what the readers throw, in case the format does not match ... this code is highly unstable, I guess.